Sources/StreamChat/Repositories/SyncRepository.swift

```diff
@@ -71,6 +74,22 @@ class SyncRepository {
     deinit {
         cancelRecoveryFlow()
     }
+    
+    // MARK: - Tracking Active
+    
+    func trackChannelListQuery(_ provider: @escaping () -> ChannelListQuery?) {
```

Comment:
I think in this case, we can move to the new provider, since there's not much knowledge of this in the current team either way. WDYT @martinmitrevski ?

Reply:
If I want to replicate this exactly, then I need to add some sort of callback as well since currently SyncRepository calls into `ChatChannelListController` and updates the `hasLoadedAllPreviousChannels`.
Then it should look like: `provider: @escaping () -> (query: ChannelListQuery, callback: ([ChatChannel]) -> Void)` (or a separate type for making this more readable).

https://github.com/GetStream/stream-chat-swift/blob/develop/Sources/StreamChat/Controllers/ChannelListController/ChannelListController.swift#L195C5-L215C6
```swift
func resetQuery(
        watchedAndSynchedChannelIds: Set<ChannelId>,
        synchedChannelIds: Set<ChannelId>,
        completion: @escaping (Result<(synchedAndWatched: [ChatChannel], unwanted: Set<ChannelId>), Error>) -> Void
    ) {
        let pageSize = query.pagination.pageSize
        worker.resetChannelsQuery(
            for: query,
            pageSize: pageSize,
            watchedAndSynchedChannelIds: watchedAndSynchedChannelIds,
            synchedChannelIds: synchedChannelIds
        ) { [weak self] result in
            switch result {
            case let .success((newChannels, unwantedCids)):
                self?.hasLoadedAllPreviousChannels = newChannels.count < pageSize
                completion(.success((newChannels, unwantedCids)))
            case let .failure(error):
                completion(.failure(error))
            }
        }
    }
```

Reply:
I played around with it a bit but still feels like it is better to do such cleanup later for the controller. It is already captured in the Notion page as well.

Reply:
I think it's best if we don't change this in the first iteration. Sync is very complex and tricky to test, so best to do as little changes as possible while we do the bigger stuff. It's anyway internal, we can change it later.